### PR TITLE
Specify the header name for the trace header

### LINKED ISSUE #704: Specify the header name for the trace header
By default, the trace header is fixed to the header `X-Runn-Trace`.

Allow header names to be specified as follows

```yaml
runners:
  myapi:
    endpoint: https://api.github.com
    trace:
      enable: true
      headername: "request-id"
```

### PR DESCRIPTION
**Pull request by k2tzumi**
close #704

### CODE REVIEW

File: runner_option.go [hunk 1]
```diff
@@ -20,11 +20,16 @@ type httpRunnerConfig struct {
 	SkipVerify           bool   `yaml:"skipVerify,omitempty"`
 	Timeout              string `yaml:"timeout,omitempty"`
 	UseCookie            *bool  `yaml:"useCookie,omitempty"`
-	Trace                *bool  `yaml:"trace,omitempty"`
+	Trace                traceConfig
```
Comment (k2tzumi) [inline runner_option.go:23]: How can I make the nested traceConfig type and the original non-nested bool type compatible with each other?

Comment (k1LoW) [reply runner_option.go:23]: Looks like we need to implement MarshalYAML as well as MarshalJSON of JSON

File: book.go [hunk 2]
```diff
@@ -260,7 +260,7 @@ func (bk *book) parseRunner(k string, v any) error {
 func (bk *book) parseHTTPRunnerWithDetailed(name string, b []byte) (bool, error) {
 	c := &httpRunnerConfig{}
 	if err := yaml.Unmarshal(b, c); err != nil {
-		return false, nil
+		return false, err
```
Comment (k1LoW) [inline book.go:263]: This place should not be err.
Look at the part where you call `parse*RunnerWithDetailed`

File: book.go [hunk 3]
```diff
@@ -323,7 +324,7 @@ func (bk *book) parseHTTPRunnerWithDetailed(name string, b []byte) (bool, error)
 func (bk *book) parseGRPCRunnerWithDetailed(name string, b []byte) (bool, error) {
 	c := &grpcRunnerConfig{}
 	if err := yaml.Unmarshal(b, c); err != nil {
-		return false, nil
+		return false, err
```
Comment (k1LoW) [inline book.go:327]: Same as https://github.com/k1LoW/runn/pull/742/files#r1439430640

File: book.go [hunk 4]
```diff
@@ -371,15 +372,17 @@ func (bk *book) parseGRPCRunnerWithDetailed(name string, b []byte) (bool, error)
...
+
 	bk.grpcRunners[name] = r
 	return true, nil
 }
 
 func (bk *book) parseDBRunnerWithDetailed(name string, b []byte) (bool, error) {
 	c := &dbRunnerConfig{}
 	if err := yaml.Unmarshal(b, c); err != nil {
-		return false, nil
+		return false, err
```
Comment (k1LoW) [inline book.go:385]: Same as https://github.com/k1LoW/runn/pull/742/files#r1439430640

File: book.go [hunk 5]
```diff
@@ -396,7 +399,7 @@ func (bk *book) parseDBRunnerWithDetailed(name string, b []byte) (bool, error) {
 func (bk *book) parseSSHRunnerWithDetailed(name string, b []byte) (bool, error) {
 	c := &sshRunnerConfig{}
 	if err := yaml.Unmarshal(b, c); err != nil {
-		return false, nil
+		return false, err
```
Comment (k1LoW) [inline book.go:402]: Same as https://github.com/k1LoW/runn/pull/742/files#r1439430640

File: book.go (hunk 2 above)
Comment (k1LoW) [reply book.go:263]: Runner is determined by whether or not Unmarshal can be done.

File: runner_option.go (hunk 1 above)
Comment (k2tzumi) [reply runner_option.go:26]: Thanks for the advice!
It took a lot of work, but I did it.
https://github.com/k1LoW/runn/pull/742/commits/597dc8c8218ee226f398133e8c29a9cbbcd84a9f

File: grpc.go [hunk 6]
```diff
@@ -788,7 +790,12 @@ func (r *grpcRequest) setTraceHeader(s *step) error {
 		return err
 	}
 	// Set Trace in the header
-	r.headers.Set("x-runn-trace", string(tj))
+	if s.grpcRunner != nil && s.grpcRunner.traceHeaderName != "" {
+		r.headers.Set(s.grpcRunner.traceHeaderName, string(tj))
+	} else {
+		// by Default
+		r.headers.Set(strings.ToLower(traceHeaderName), string(tj))
```
Comment (k1LoW) [inline grpc.go:797]:
```suggestion
		r.headers.Set(traceHeaderName, string(tj))
```

Lowercased internally.

https://github.com/grpc/grpc-go/blob/dbbcf59957fec0bd58063224cbf105b3b3698d4e/metadata/metadata.go#L116

File: trace.go [hunk 7]
```diff
@@ -1,5 +1,7 @@
 package runn
 
+const traceHeaderName = "X-Runn-Trace"
```
Comment (k1LoW) [inline trace.go:3]: This constant is the default value, correct?

```suggestion
const defaultTraceHeaderName = "X-Runn-Trace"
```

File: runner_option.go [hunk 8]
```diff
@@ -327,3 +335,21 @@ func LocalForward(l string) sshRunnerOption {
 		return nil
 	}
 }
+
+func (t *traceConfig) UnmarshalYAML(b []byte) error {
```
Comment (k1LoW) [inline runner_option.go:339]: GREAT! 👍

File: runner_option.go [hunk 9]
```diff
@@ -20,11 +23,16 @@ type httpRunnerConfig struct {
...
 	UseCookie            *bool  `yaml:"useCookie,omitempty"`
-	Trace                *bool  `yaml:"trace,omitempty"`
+	Trace                traceConfig
 
 	openApi3Doc *openapi3.T
 }
 
+type traceConfig struct {
+	Enable     *bool  `yaml:"enable,omitempty"`
+	Headername string `yaml:"headername,omitempty"`
```
Comment (k1LoW) [inline runner_option.go:33]: Is there any reason why it shouldn't be `headerName`?

```suggestion
	HeaderName string `yaml:"headerName,omitempty"`
```

File: runner_option.go [hunk 10]
```diff
@@ -327,3 +335,21 @@ func LocalForward(l string) sshRunnerOption {
...
 	}
 }
+
+func (t *traceConfig) UnmarshalYAML(b []byte) error {
+	if enable, err := strconv.ParseBool(strings.TrimSpace(string(b))); err == nil {
+		t.Enable = &enable
+		return nil
+	}
+
+	type TC traceConfig
```
Comment (k1LoW) [inline runner_option.go:345]: Does it have to be an uppercase variable?

I thought I could make `traceConfig` into `TraceConfig` if needed.

**Comment by k1LoW** [comment]
NICE!

File: runner_option.go (hunk 10 above)
Comment (k2tzumi) [reply runner_option.go]: I've tried to refer to this one.
https://github.com/k1LoW/octocov/blob/main/report/json.go

File: option.go [hunk 11]
```diff
@@ -401,9 +402,10 @@ func GrpcRunner(name string, cc *grpc.ClientConn) Option {
 		}
 		delete(bk.runnerErrs, name)
 		r := &grpcRunner{
-			name: name,
-			cc:   cc,
-			mds:  map[string]protoreflect.MethodDescriptor{},
+			name:            name,
+			cc:              cc,
+			mds:             map[string]protoreflect.MethodDescriptor{},
+			traceHeaderName: strings.ToLower(defaultTraceHeaderName),
```
Comment (k1LoW) [inline option.go:408]: Same as https://github.com/k1LoW/runn/pull/742#discussion_r1441145862

```suggestion
			traceHeaderName: defaultTraceHeaderName,
```

File: option.go [hunk 12]
```diff
@@ -418,9 +420,10 @@ func GrpcRunnerWithOptions(name, target string, opts ...grpcRunnerOption) Option
 		}
 		delete(bk.runnerErrs, name)
 		r := &grpcRunner{
-			name:   name,
-			target: target,
-			mds:    map[string]protoreflect.MethodDescriptor{},
+			name:            name,
+			target:          target,
+			mds:             map[string]protoreflect.MethodDescriptor{},
+			traceHeaderName: strings.ToLower(defaultTraceHeaderName),
```
Comment (k1LoW) [inline option.go:426]: Same as https://github.com/k1LoW/runn/pull/742#discussion_r1441145862

```suggestion
			traceHeaderName: defaultTraceHeaderName,
```

**Review by k1LoW — COMMENTED** [review]
Two reviews were omitted 💦

**Comment by k1LoW** [comment]
@k2tzumi 
Everything else is fine! Thank you!!!

**Pull request merged**